### PR TITLE
Add Academy/ASC Common LUT Format (CLF) read support (#36)

### PR DESCRIPTION
**Pull request by BernardLefebvre**
This branch adds full read support for the Academy/ASC Common LUT Format (CLF), Academy S-2014-006, along with new and improved processing ops.

New Range and ASC CDL ops have been added along with major enhancements to the Lut1D and Lut3D ops:  
-- The Range op provides easy level conversions and clamping.  
-- The ASC CDL op was added to allow implementation of various commonly used CDL formulas including one that is now faithful to the official ASC 1.2 spec (which OCIO's original implementation did not match).
-- Lut1D now supports the half-domain feature of CLF which allows a 1D LUT to be built using the same domain as a half-precision floating-point number.
-- Lut3D now supports inversion, meaning that all ops and color space conversions may now be applied in either direction.  
-- By default, the LUT inversion is using the fast approximation. Later a flag would select the fast or exact (but slow) LUT inversion algorithm.

This branch introduces an OpData layer containing the detailed operator machinery and unit tests and which allowed us to avoid making the core/*Op modules too large.   This code is in a new core/opdata directory.

The CPU renderers for the Lut1D, Lut3D, and matrix ops is now using new more optimized code.  This is now in a new core/cpu directory rather than in the core/*Op modules.  Support for integer bit-depths is still in progress and will be completed in a follow-on branch.  In the current branch, INTERP_NEAREST for Lut1D and Lut3D is temporarily handled as linear.

A core/clf directory was added containing the parsing code for the CLF format.  This makes use of the Expat XML parser which is faster and allows handling large files with a smaller memory footprint than TinyXML.

Many unit tests were added, including tests that validate the GPU renderer against the CPU renderer.

The ocioconvert and ociodisplay apps were updated:
-- ocioconvert now allows rendering with the GPU when compiled with the BUILD_WITH_GPU flag.
-- ociodisplay now allows use of the new GPU renderer.

The half module from ILMBase and the Expat library have been added under ext and the LICENSE file updated.

### CODE REVIEW

**Comment by hodoulp** [comment]
If you have some time please review to the pull request because that work is extremely important for the next steps of OCIO v2. Any kind of review is welcome, it could be to only look at a specific part of the code, the documentation, the unit tests...

**Comment by hodoulp** [comment]
Thanks for reviewing the pull request. As mentioned the work represents the complete implementation of the CLF reader as well as the foundation for later improvements such as the CTF reader. A complete implementation means that lot of files are updated. As a reminder, only implementing a small subset of an ICC Profile (without any new ops) already changed 50 files.
We deeply apologize to provide such a huge pull request. In our will to provide a complete and useable feature we missed some intermediate self-contained changes that could have be reviewed independently from the CLF pull request.
To keep track of the feature, we will update that pull request with comments and answers and will soon submit other pull requests.

**Comment by hodoulp** [comment]
Please check the pull request #564 to have smaller part of that work.

**Comment by hodoulp** [comment]
Here is the pull request #565 to enhance ocioconvert for testing.

File: src/core/ExponentOps.cpp [hunk 1]
```diff
@@ -72,7 +72,13 @@ OCIO_NAMESPACE_ENTER
             virtual std::string getInfo() const;
             virtual std::string getCacheID() const;
             
+            virtual BitDepth getInputBitDepth() const;
```
[the remark this answers is not in the document]
Comment (hodoulp) [reply src/core/ExponentOps.cpp]: As the CLF file format does not support the Log & Exponent ops we did not implement the input and output bit depths. Coming pull request around the CTF file format will finalize the feature.

File: CMakeLists.txt [hunk 2]
```diff
@@ -246,7 +250,56 @@ else(USE_EXTERNAL_TINYXML)
     endif()
     set_target_properties(TINYXML_LIB PROPERTIES FOLDER External)
 endif(USE_EXTERNAL_TINYXML)
-    
+
+###############################################################################
+### libexpat ###
```
[the remark this answers is not in the document]
Comment (hodoulp) [reply CMakeLists.txt]: As explained by Bernard (in one of the  OCIO Slack channel), CLF file format parsing using tinyXML is consuming a lot of memory (explaining why synColor is using the eXpact library). However, eXpact implementation requires much more work. The complete removal of the tinyXML will done in the following pull request(s) (we hope with the help of the community) in order to control the size of this pull request.

File: CMakeLists.txt [hunk 3]
```diff
@@ -369,6 +422,18 @@ if(YAML_CPP_VERSION VERSION_LESS "0.5.0")
     set(YAML_CPP_COMPILE_FLAGS "-DOLDYAML")
 endif()
 
+###############################################################################
+### ilmbase half ###
```
[the remark this answers is not in the document]
Comment (hodoulp) [reply CMakeLists.txt]: Good point.

**Comment by hodoulp** [comment]
Concerning the complete support of 'not SEE' apply, we planned to do it in following pull request. For the short-term we will improve the pull request to better highlight the current limitations.

**Comment by hodoulp** [comment]
The pull request #566 is adding the major & minor versions to the configuration file.

**Comment by BernardLefebvre** [comment]
Moved existing files comments changes (LUT case, spelling issues) to https://github.com/imageworks/OpenColorIO/pull/569

**Comment by hodoulp** [comment]
#567 & #570 are also part of the effort.

**Comment by hodoulp** [comment]
We had a meeting with Mark Boorer (@Shootfast) and Sean Cooper (@slooper) to discuss their feedback about the CLF PR.  We really appreciate their taking the time to review our work and share their knowledge of the OCIO project.  We agreed to try and split up PR #556 into a series of smaller PRs (although even after we've broken it down as much as is feasible, one or more may still be quite large).  However we will try to constrain the scope of the PRs to a specific change, rather than having everything needed to implement a whole feature.  Although this represents a lot of extra work for us, we agree it will make the PRs/commits easier to follow, both now and for future maintainers.

So based on the discussion, here is our plan for the PRs we are thinking of doing in order to get us to the CLF feature in a more agreeable manner:
1.	**[ASAP]** Merge pending pull requests nos. #565 , #567 , #568 , #569 , and #570 
a.	Must be first to help task 2
2.	**[ASAP]** Refactor the project structure (to be done by Mark Boorer)
a.	Based on the agreement on the file/directory structure
b.	Get community feedback
3.	Switch from tinyXML to eXpat
a.	Based on the agreement to switch the XML library upfront 
b.	Even if the justification will come later (i.e. CLF/CTF reader)
4.	[Proposal] Support of bit depths: (8-12-16) integer & half & float, for RGBA and RGB
a.	Slack publish & discussion, and perhaps a skeleton in a pull request
5.	Refactor & enhance existing Ops
a.	Agree on new design which is more powerful but also easy to read/understand
b.	Add new functionality from PR #556
6.	Add support of bit depths
a.	Add a complete implementation for all ops
b.	Enhance public API apply() method with bit depth capabilities
7.	Add Range op from the CLF specification
8.	Add CDL op to also support the CLF specification
a.	While preserving existing implementation for OCIO v1 configuration files
9.	Add CLF file format reader

**Comment by hodoulp** [comment]
Added #575 & #576

**Comment by scoopxyz** [comment]
This thread is so long, I don't know where we stand. Is this PR good for closing? I haven't seen an additional PR for CLF/CTF since the restructure

**Comment by hodoulp** [comment]
All the current ADSK PR's are related to the CLF support. As discussed, we are decomposing this PR in smaller ones which should take some time to have all the needed PRs to fully implement the CLF support. And many recently logged issues are related to the file restructuration.

Anyway, as this pull request will never be submitted, I close the PR.

**Pull request closed without merging**